### PR TITLE
Oceanwater 457 refactor thresholding algo to scan samples once

### PR DESCRIPTION
**Pull request by Samahu**
## Summary of Changes:
This PR improves and addresses several issues in current implementation/integration of the ground detection:
The improvements include:
- Scan velocity samples only once
- Catch peaks that could have been missed
- Make sure that the peak detector state is reset on every invocation (to enable invoking multiple times in one session).
- Limit resource usage

## Verify:
Launch the simulation using
```bash
roslaunch ow europa_terminator_workspace.launch
```
Then plan a guarded move and execute it.
Try Invoking the plan multiple times in one session.
Report how to often do you successfully get the ground detected signal. This doesn't address making the ground detection robust yet. See the note down below. 

## Objectives
- Verify functionality described
- Verify that this change didn't break any existing functionality (for example the **guarded_move_result**)
- Review code and spot an potential issues or improvements

**Linked Issue:** https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-457

__Note:__ This PR doesn't include the necessary changes to make robust against minor tweaks to the simulation. This will be addressed as part of issue: https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-461

### CODE REVIEW

**Comment by kmdalal** [comment]
I've tested using Rviz only.  I tried 3 GuardedMoves in the same roslaunch mentioned above.  I noticed the following:

- slight arm shakes each time, of short duration, and it stops
-  while planning the ghost arm moves well below the RViz plane.  However, when the trajectory is published, the solid arm stops well _above_ the RViz plane.  I'm guessing this is because the terrain surface is actually above the plane.
- ground is found each time, with slightly different values:

position: 
  x: 2.00554227829
  y: -0.0919659584761
  z: 0.0991680249572
frame: "base_link"
---
success: True
position: 
  x: 2.01186013222
  y: -0.0920400843024
  z: 0.0513819232583
frame: "base_link"
---
success: True
position: 
  x: 1.9993981123
  y: -0.0919873639941
  z: 0.13308057189
frame: "base_link"

File: ow_lander/scripts/trajectory_publisher.py [hunk 1]
```diff
@@ -141,28 +116,30 @@ def talker(req):
     guarded_move_pub = rospy.Publisher('/guarded_move_result', GuardedMoveResult, queue_size=10)
 
     # Start subscriber for the joint_states
+    time.sleep(2) # subscribe right after coming from sleep .. this way we maintain values when moving downwards
```
Comment (mogumbo) [inline ow_lander/scripts/trajectory_publisher.py:119]: I don't understand this comment. Can you clarify it?

File: ow_lander/scripts/peak_detection_real_time.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    Use for every new session without having to recreate the object. """
+    self.counter = 0
+    self.y_raw = 0
+    self.y_filtered = 0
+    self.signal = 0
+    self.avg_filter = SlidingWindow(self.lag, np.mean)
+    self.std_filter = SlidingWindow(self.lag, np.std)
+
+
+  def _warmup(self):
```
Comment (mogumbo) [inline ow_lander/scripts/peak_detection_real_time.py:58]: I don't understand what this function is for or what "warmup" means in this context. Can you add a comment explaining what it is for?

**Review by mogumbo — COMMENTED** [review]
I ran guarded move 6 times and got 3 accurate detections and 3 premature detections.

File: ow_lander/scripts/peak_detection_real_time.py (hunk 2 above)
Comment (Samahu) [reply ow_lander/scripts/peak_detection_real_time.py:58]:
```suggestion
  def _warmup(self):
  """
  accumulate enough samples (determined by lag) before performing detection 
  :return: always returns zero. i.e. no peak or anomaly.
  """
```

File: ow_lander/scripts/trajectory_publisher.py (hunk 1 above)
Comment (Samahu) [reply ow_lander/scripts/trajectory_publisher.py:119]:
```suggestion
    time.sleep(2) # wait for 2 seconds till the arm settles 
```

File: ow_lander/scripts/trajectory_publisher.py [hunk 3]
```diff
@@ -141,28 +116,30 @@ def talker(req):
...
+    global ground_detected, peak_detection
```
Comment (Samahu) [inline ow_lander/scripts/trajectory_publisher.py:120]:
```suggestion
    # begin tracking velocity values as soon as the scoop moves downward 
    global ground_detected, peak_detection
```

**Review by damianacatanoso — COMMENTED** [review]
Every time the scoop touches the ground, it leaves some holes. I did these tests ignoring this.
- Attempt 1: 
Ground successfully found
success: True
position:  
 x: 2.01223421097
 y: -0.0922526419163
 z: -0.138023853302
frame: "base_link"

- Attempt 2:
Ground successfully found
success: True
position:  
 x: 2.04994916916
 y: -0.0960469692945
 z: -0.0609309785068
frame: "base_link"

- Attempt 3:
Success
success: True
position:  
 x: 2.06120491028
 y: -0.096640214324
 z: -0.0393313355744
frame: "base_link"

- Attempt 4:
Success
success: True
position:  
 x: 2.05466198921
 y: -0.096353366971
 z: -0.052013926208
frame: "base_link"

- Attempt 5:
False detection
success: True
position:  
 x: 1.99600422382
 y: -0.0920018553734
 z: 0.149415940046
frame: "base_link"

- Attempt 6:
False detection
success: True
position:  
 x: 1.99928450584
 y: -0.0919585898519
 z: 0.133393034339
frame: "base_link"

- Attempt 7:
False detection
success: True
position:  
 x: 2.00002598763
 y: -0.0919710621238
 z: 0.13118493557
frame: "base_link"

**Comment by mogumbo** [comment]
I approve this PR, though I can't find the approval button like I found on previous PRs. GitHub PRs are pretty confusing.

**Pull request merged**